### PR TITLE
Data gatherer

### PR DESCRIPTION
**Pull request by BeckySharp**
Let's discuss

### CODE REVIEW

**Comment by myedibleenso** [comment]
>Let's discuss

@BeckySharp , can you please summarize this?

**Comment by BeckySharp** [comment]
yes  @myedibleenso for sure!  

The overarching goal is to (across a few PRs) make the Mentions much easier to work with _after_ extracting.  
OK, so the idea is that a "datagatherer" is the guy who is able to get the annotation content of the mentions (e.g., raw, lemmas, tags, entities, etc.).. In this sense, the ExtractorEngine is no longer responsible for this directly (clarifying "responsibility").

Going right along with this -- the idea is that initially, extracted mentions are *minimal* (like they currently are), they don't know their text, their containing sentence text, their linguistic annotations etc.  Then, if you want to (as many downstream people do) you can `populate` them -- basically *add this content* so that the Mentions themselves can tell you things you want to know about them (text etc.).  To try and maintain the efficiency of things, there are different levels of population -- (a) none (i.e., don't populate), (b) `display` (displayField only, for both the mention and the containing document/sentence), and (c) `all`, (all stored fields available in the index for both the mention and the containing document/sentence).

This PR has the decoupling, has the attributes of the Mention, and has basic access methods in the Mention (`mention.text`), but in a _later_ PR I plan to add more access methods to add some of the much-loved ease of use of Odin Mentions.

The decision to use `private var` in the Mention was to avoid creating extra objects (and so to make it faster/lighter weight), however, that behavior can be modified.  If not modified, fwiw, from the outside they behave as immutable.

Please ask more questions, give comments, etc

File: core/src/main/scala/ai/lum/odinson/DataGatherer.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  def getTokens(scoreDoc: OdinsonScoreDoc): Array[String] = {
+    getTokens(scoreDoc.doc, displayField)
+  }
+
+  def getTokens(scoreDoc: OdinsonScoreDoc, fieldName: String): Array[String] = {
+    getTokens(scoreDoc.doc, fieldName)
+  }
+
+  def getTokens(docID: Int, fieldName: String): Array[String] = {
+    TokenStreamUtils.getTokens(docID, fieldName, indexSearcher, analyzer)
```
Comment (BeckySharp) [inline core/src/main/scala/ai/lum/odinson/DataGatherer.scala:92]: per marco -- try to intelligently load only fields we need:
https://lucene.apache.org/core/6_6_6/core/index.html

**Comment by BeckySharp** [comment]
`val c = b.map{x => x.append(2);x}`

**Comment by BeckySharp** [comment]
With this PR now -- Mentions can populate themselves with their lexical info (they have a pointer to a DataGatherer).
 --> `Mention.populateFields`
This includes the lexical info for the mention itself as well as the Document (i.e., sentence)

Also, the `ExtractorEngine` has a method for getting mentions called `extractAndPopulate` which returns an Iterator of mentions that are already populated.  This makes it one less step for the consumer if they want that.

Per request, TokenStreamUtils was modified to only fetch the fields desired from the index to save time.

Tests are added in TestMention and TestTokenStreamUtils

in an upcoming PR I plan to deprecate some `getTokens` methods in the ExtractorEngine and DataGatherer.

**Comment by BeckySharp** [comment]
fyi @myedibleenso and @danebell -- this may make things easier in the backend?  If adjustments are needed for your purposes also, please let me know! 
thanks!

File: core/src/main/scala/ai/lum/odinson/lucene/analysis/TokenStreamUtils.scala [hunk 2]
```diff
@@ -2,23 +2,41 @@ package ai.lum.odinson.lucene.analysis
...
+  def getTokensFromMultipleFields(
     docID: Int,
-    fieldName: String,
+    fieldNames: Set[String],
     indexSearcher: IndexSearcher,
     analyzer: Analyzer
-  ): Array[String] = {
-    val doc = indexSearcher.doc(docID)
+  ): Map[String, Array[String]] = {
+    val doc = getDoc(docID, fieldNames, indexSearcher)
```
Comment (BeckySharp) [inline core/src/main/scala/ai/lum/odinson/lucene/analysis/TokenStreamUtils.scala:27]: just need a reader? if so, only use reader above in the datagatherer

**Review by marcovzla — APPROVED** [review]
Thanks!

**Pull request merged**